lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2600,19 +2600,86 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
   AutoDiffParameterIndices *checkedWrtParamIndices =
       attr->getParameterIndices();
 
+  // Predicate checking if a type has associated tangent and cotangent spaces.
+  auto hasAssociatedSpaces = [&](Type type) -> bool {
+    return (bool)type->getAutoDiffAssociatedVectorSpace(
+               AutoDiffAssociatedVectorSpaceKind::Tangent, lookupConformance) &&
+           (bool)type->getAutoDiffAssociatedVectorSpace(
+               AutoDiffAssociatedVectorSpaceKind::Cotangent, lookupConformance);
+  };
+
   // If checked wrt param indices are not specified, compute them using parsed
   // wrt param indices.
   if (!checkedWrtParamIndices) {
     AutoDiffParameterIndicesBuilder autoDiffParameterIndicesBuilder(
         originalFnTy);
     if (parsedWrtParams.empty()) {
+      // This is odd because Self is placed at the end (Args..., Self) due
+      // to existing currying logic. Unfortunately we want to talk about the
+      // underlying method which actually is (Args..., Self), but this function
+      // is (Self) -> (Args...) -> R, but we don't want to talk about normal
+      // curried functions.
+      // TODO: Clean all this up.
+      SmallVector<Type, 4> allWRTParamTypes;
       if (original->isStatic() || isa<ConstructorDecl>(original)) {
         auto *methodTy =
             original->getMethodInterfaceType()->castTo<AnyFunctionType>();
-        autoDiffParameterIndicesBuilder
-            .setParameters(0, methodTy->getNumParams());
+        if (methodTy->getResult()->getAs<AnyFunctionType>() != nullptr) {
```

Comment:
Please replace occurrences of `getAs<AnyFunctionTypee>() != nullptr` with `is<AnyFunctionType>()`.
```suggestion
        if (methodTy->getResult()->is<AnyFunctionType>()) {
```